stdlib/sql/open.go

```diff
@@ -0,0 +1,22 @@
+package sql
+
+import (
+	"database/sql"
+)
+
+type openFunc func() (*sql.DB, error)
+
+func defaultOpenFunction(driverName, dataSourceName string) openFunc {
+	return func() (*sql.DB, error) {
+		return sql.Open(driverName, dataSourceName)
+	}
+}
+
+func getOpenFunc(driverName, dataSourceName string) openFunc {
+	switch driverName {
+	case "mssql", "sqlserver":
+		return mssqlOpenFunction(driverName, dataSourceName)
```

Comment:
Can you explain why you need to use a special open function here? It looks like `github.com/denisenkom/go-mssqldb` calls `sql.Register` for `mssql` and `sqlserver` so I would have thought that stdlib `sql.Open` would just work.

Reply:
In case when Azure authentication is used, it is necessary to call `sql.Open(connector)` instead of `sql.Open(driverName, dsn)`, because the connector has access to access token. The instance of `driver.Connector` is created using `NewAccessTokenConnector()` from `github.com/denisenkom/go-mssqldb`, which takes a `dsn` and a function that returns OAuth token as parameters and return the connector.

The documentation on Azure auth is rather brief, [Azure Active Directory authentication - preview](https://github.com/denisenkom/go-mssqldb#azure-active-directory-authentication---preview). It show the following example:
```
conn, err := mssql.NewAccessTokenConnector(
  "Server=test.database.windows.net;Database=testdb",
  tokenProvider)
if err != nil {
	// handle errors in DSN
}
db := sql.OpenDB(conn)
```

I could not figure it out otherwise other than introducing this `openFunc`. So if user does not request Azure authentication for SQL server, the `mssqOpenFunc` default to `sql.Open`, otherwise it returns something like the above.

There is also an [example](https://github.com/denisenkom/go-mssqldb/tree/master/examples/azuread-accesstoken) of access token connector .

The `github.com/denisenkom/go-mssqldb` does not reference Azure `go-autorest` module so it my understanding that the actual authentication and providing `tokenProvider` function to the connector instance is left up to `go-mssqldb` user.

Reply:
Sorry for the delay in responding.

Could you add your answer above to a comment in the code, so that if we ever need to work on this code in the future, we have some context about why you had to do it this way?

Thanks for this great work. We'd like to get one more person on our team to review this code before you merge, I'll coordinate that today.

Reply:
Code comment added, thank you.